src/serac/physics/thermal_conduction_functional.hpp

```diff
@@ -0,0 +1,481 @@
+            TimesteppingOptions{TimestepMethod::BackwardEuler, DirichletEnforcementMethod::RateControl}};
+  }
+
+  /**
+   * @brief Construct a new Thermal Solver object
+   *
+   * @param[in] options The system solver parameters
+   * @param[in] name An optional name for the physics module instance
+   */
+  ThermalConductionFunctional(const SolverOptions& options, const std::string& name = "")
+      : BasePhysics(1, order),
+        temperature_(
+            StateManager::newState(FiniteElementState::Options{.order      = order,
+                                                               .vector_dim = 1,
+                                                               .ordering   = mfem::Ordering::byNODES,
+                                                               .name       = detail::addPrefix(name, "temperature")})),
+        M_functional_(&temperature_.space(), &temperature_.space()),
+        K_functional_(&temperature_.space(), &temperature_.space()),
+        residual_(temperature_.space().TrueVSize()),
+        ode_(temperature_.space().TrueVSize(), {.u = u_, .dt = dt_, .du_dt = previous_, .previous_dt = previous_dt_},
+             nonlin_solver_, bcs_)
+  {
+    SLIC_ERROR_ROOT_IF(mesh_.Dimension() != dim,
+                       fmt::format("Compile time dimension and runtime mesh dimension mismatch"));
+
+    state_.push_back(temperature_);
+
+    nonlin_solver_ = mfem_ext::EquationSolver(mesh_.GetComm(), options.T_lin_options, options.T_nonlin_options);
+    nonlin_solver_.SetOperator(residual_);
+
+    // Check for dynamic mode
+    if (options.dyn_options) {
+      ode_.SetTimestepper(options.dyn_options->timestepper);
+      ode_.SetEnforcementMethod(options.dyn_options->enforcement_method);
+      is_quasistatic_ = false;
+    } else {
+      is_quasistatic_ = true;
+    }
+
+    dt_          = 0.0;
+    previous_dt_ = -1.0;
+
+    int true_size = temperature_.space().TrueVSize();
+    u_.SetSize(true_size);
+    previous_.SetSize(true_size);
+    previous_ = 0.0;
+
+    zero_.SetSize(true_size);
+    zero_ = 0.0;
+  }
+
+  /**
+   * @brief Set essential temperature boundary conditions (strongly enforced)
+   *
+   * @param[in] temp_bdr The boundary attributes on which to enforce a temperature
+   * @param[in] temp_bdr_coef The prescribed boundary temperature
+   */
+  template <typename TemperatureType>
+  void setTemperatureBCs(const std::set<int>& temp_bdr, TemperatureType temp)
+  {
+    // Project the coefficient onto the grid function
+    temp_bdr_coef_ = std::make_shared<mfem::FunctionCoefficient>([temp](const mfem::Vector& x, double t) -> double {
+      tensor<double, dim> x_tensor;
+      x_tensor[0] = x[0];
+      if constexpr (dim > 1) {
+        x_tensor[1] = x[1];
+      }
+      if constexpr (dim > 2) {
+        x_tensor[2] = x[2];
+      }
+
+      return temp(x_tensor, t);
+    });
+
+    bcs_.addEssential(temp_bdr, temp_bdr_coef_, temperature_);
+  }
```

Comment:
It seems more natural for this function to accept an "`mfem::Coefficient`" or a `std::function < double(const mfem::Vector&, double) >` for its second argument.

Reply:
Yep, I think when we allow for arguments of arbitrary type it might also be worth introducing corresponding traits classes/static_asserts so we can provide reasonable compile error messages if the user screws something up.

Reply:
I'm trying to avoid `Coefficients` whenever possible because the ownership semantics are weird. I think std::function makes sense.